R/eval-walk.R

```diff
@@ -286,12 +286,23 @@ call_kind <- function(expr, error_call) {
   fn <- as_string(head)
 
   if (fn %in% c("$", "[[") && identical(node_cadr(expr), quote(.data))) {
-    lifecycle::deprecate_warn("1.2.0",
-      what = I("Use of .data in tidyselect expressions"),
-      with = I("`any_of()` or `all_of()`")
-    )
-
     validate_dot_data(expr, error_call)
+
+    what <- I("Use of .data in tidyselect expressions")
+    if (fn == "$") {
+      var <- as.character(expr[[3]])
+      str <- encodeString(var, quote = '"')
+
+      lifecycle::deprecate_warn("1.2.0", what,
+        details = glue("Please use `{str}` instead of `data${var}`")
+      )
+    } else if (fn == "[[") {
+      var <- expr_deparse(expr[[3]])
```

Comment:
This could be multiline, for instance if a confused user wrote `.data[[{{ foo }}]]`. Use `as_label()` instead?